p2p/nodechecker_test.go

```diff
@@ -0,0 +1,241 @@
+			time.Sleep(100 * time.Millisecond)
+
+			nodeB.host.discovery.CheckPeers(context.Background())
+			time.Sleep(100 * time.Millisecond)
+
+			addressBookAddresses := nodeB.host.discovery.GetAddresses()
+			if len(testCase.expected) == 2 {
+				require.True(t, len(addressBookAddresses) > 0, "nodeB should have at least 1 address")
+			} else {
+				peerStoreAddressList := nodeB.host.Peerstore().Addrs(nodeA.host.ID())
+				require.Equal(t, len(testCase.expected), len(peerStoreAddressList), "nodeB should have %d addresses", len(testCase.expected))
+				require.Equal(t, len(testCase.expected), len(addressBookAddresses), "nodeB should have %d addresses", len(testCase.expected))
+				for _, addr := range testCase.expected {
+					var expectAddress string
+					switch addr {
+					case oldAddr:
+						expectAddress = addrAOld.String()
+					case newAddr:
+						expectAddress = addrANew.String()
+					}
+					require.Equal(t, expectAddress, peerStoreAddressList[0].String(), "nodeB should have correct address of nodeA")
+					require.Equal(t, expectAddress+"/p2p/"+nodeA.host.ID().String(), addressBookAddresses[0].String(), "nodeB should have correct address of nodeA")
+				}
+			}
+		})
+	}
+}
+
+func generateAddressV4(t *testing.T) ma.Multiaddr {
+	blackholeIP6 := net.ParseIP("100::")
+	sk, err := p2putil.RandTestBogusPrivateKey()
+	require.NoError(t, err)
+
+	id, err := peer.IDFromPrivateKey(sk)
+	require.NoError(t, err)
+
+	suffix := id
+	if len(id) > 8 {
+		suffix = id[len(id)-8:]
+	}
+	ip := append(net.IP{}, blackholeIP6...)
+	copy(ip[net.IPv6len-len(suffix):], suffix)
+	port, err := getFreePort()
+	require.NoError(t, err)
+
+	a, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", port))
+	require.NoError(t, err)
+	return a
+}
+
+func getFreePort() (int, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return 0, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+func generateNode(t *testing.T, addr ma.Multiaddr, bootNode string) *hostWrapper {
+	addrFactory := &addressFactory{
+		blockAddressesMu: &sync.RWMutex{},
+		blockAddresses:   []string{},
+		mark:             make(chan struct{}),
+	}
+
+	node, err := libp2p.New(context.Background(),
+		libp2p.ListenAddrs(addr),
+		libp2p.DisableRelay(),
+		libp2p.NATPortMap(),
+		libp2p.AddrsFactory(addrFactory.addressFactory),
+	)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, node.Close())
+	})
+	cnf := DefaultConfig()
+	cnf.CheckPeersUsedBefore = time.Millisecond * 1
+	if bootNode != "" {
+		cnf.Bootnodes = append(cnf.Bootnodes, bootNode)
+	}
+	host, err := Upgrade(node, WithConfig(cnf))
+	require.NoError(t, err)
+	return &hostWrapper{
+		host:      host,
+		transport: addrFactory,
+	}
+}
+
+func discoveryBootstrap(disc *peerexchange.Discovery) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
why it needs defer cancel() if function exits only if Bootstrap is canceled?

Reply:
removed this